File: basic_dsa_two_hundred_fifty_two.cpp
/**
 * Basic DSA - Height Checker (LeetCode 1051)
 * Return number of indices where heights[i] != expected[i].
 */
#include <iostream>
#include <vector>
#include <algorithm>

int heightChecker(std::vector<int>& heights) {
    std::vector<int> expected = heights;
    std::sort(expected.begin(), expected.end());
    int cnt = 0;
    for (int i = 0; i < (int)heights.size(); i++)
        if (heights[i] != expected[i]) cnt++;
    return cnt;
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Height Checker\n";
    std::cout << std::string(60, '=') << "\n";

    std::vector<int> v1 = {1, 1, 4, 2, 1, 3};
    std::cout << "[1,1,4,2,1,3] -> " << heightChecker(v1) << "\n";

    std::vector<int> v2 = {5, 1, 2, 3, 4};
    std::cout << "[5,1,2,3,4] -> " << heightChecker(v2) << "\n";

    std::vector<int> v3 = {1, 2, 3, 4, 5};
    std::cout << "[1,2,3,4,5] -> " << heightChecker(v3) << "\n";
}

File: basic_dsa_two_hundred_fifty_two.py
"""
Basic DSA - Jump Game II
You are given an integer array nums where nums[i] is the maximum jump length from index i.
Return the minimum number of jumps to reach the last index.
Assume you can always reach the last index.
"""


def jump(nums):
    """
    Greedy BFS-level approach:
    Track current range [0..end] reachable with 'steps' jumps, and the farthest reach
    within that range. When i hits end, we must take another jump.
    Time Complexity: O(n)
    Space Complexity: O(1)
    """
    if len(nums) <= 1:
        return 0

    steps = 0
    end = 0
    far = 0
    for i in range(len(nums) - 1):
        far = max(far, i + nums[i])
        if i == end:
            steps += 1
            end = far
    return steps


if __name__ == "__main__":
    print("=" * 60)
    print("Problem: Jump Game II")
    print("=" * 60)

    print(jump([2, 3, 1, 1, 4]))  # 2
    print(jump([2, 3, 0, 1, 4]))  # 2
